stdlib/asm/crypto/hashes/blake3.masm

```diff
@@ -1,579 +1,447 @@
+# Given blake3 state matrix on stack top ( in order ) as 16 elements ( each of 32 -bit ),
+# this routine computes output chaining value i.e. 2-to-1 hashing digest.
 #
-# see https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L116-L119
+# Expected stack state:
+#
+# [state0, state1, state2, state3, state4, state5, state6, state7, state8, state9, state10, state11, state12, state13, state14, state15]
+#
+# After finalizing, stack should look like
+#
+# [dig0, dig1, dig2, dig3, dig4, dig5, dig6, dig7]
+#
+# See https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L116-L119 ,
 # you'll notice I've skipped executing second statement in loop body of above hyperlinked implementation,
-# that's because it doesn't dictate what output of 2-to-1 hash will be
-proc.prepare_digest.0
-    dup.8
+# that's because it doesn't dictate what output of 2-to-1 hash will be.
+proc.finalize
+    movup.8
     u32xor
 
-    dup.9
-    movup.2
+    swap
+    movup.8
     u32xor
     swap
 
-    dup.10
-    movup.3
+    movup.2
+    movup.8
     u32xor
     movdn.2
 
-    dup.11
-    movup.4
+    movup.3
+    movup.8
     u32xor
     movdn.3
 
-    dup.12
-    movup.5
+    movup.4
+    movup.8
     u32xor
     movdn.4
 
-    dup.13
-    movup.6
+    movup.5
+    movup.8
     u32xor
     movdn.5
 
-    dup.14
-    movup.7
+    movup.6
+    movup.8
     u32xor
     movdn.6
 
-    dup.15
+    movup.7
     movup.8
     u32xor
     movdn.7
 end
 
-# column-wise mixing
-# see https://github.com/BLAKE3-team/BLAKE3/blob/da4c792d8094f35c05c41c9aeb5dfe4aa67ca1ac/reference_impl/reference_impl.rs#L55-L59
+# Given blake3 state matrix ( total 16 elements, each of 32 -bit ) and 
+# 8 message words ( each of 32 -bit ), this routine performs column-wise mixing
+# of message words into blake3 hash state.
+#
+# Functionality wise this routine is equivalent to https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L55-L59
+#
+# Expected stack state:
+#
+# [state0_3_addr, state4_7_addr, state8_11_addr, state12_15_addr, m0, m1, m2, m3, m4, m5, m6, m7]
+#
+# Note, state_`i`_`j`_addr -> absolute address of {state[i], state[i+1], state[i+2], state[i+3]} in memory | j = i+3
+#
+# Meaning four consecutive blake3 state words can be read from memory easily.
+#
+# Final stack state:
+#
+# [state0, state1, state2, state3, state4, state5, state6, state7, state8, state9, state10, state11, state12, state13, state14, state15]
+#
+# i.e. whole blake3 state is placed on stack ( in order ).
 proc.columnar_mixing.1
-    pushw.mem
+    swapw.2
+    swapw
+
+    movup.7
+    movup.6
+    movup.5
+    movup.4
+
     popw.local.0
 
-    pushw.local.0
-    drop
-    repeat.2
-        swap
-        drop
-    end
-    pushw.mem           # = b
+    movup.4
+    movup.5
 
-    pushw.local.0
-    repeat.3
-        swap
-        drop
-    end
-    pushw.mem           # = a
+    pushw.mem
+    movup.4
+    pushw.mem
```

Comment:
You might be able to replace the sequence
```
popw.local.0
movup.4
movup.5
pushw.mem
movup4
pushw.mem
```
With something like
```
storew.local.0
movup.8
loadw.mem
movup.8
pushw.mem
```
If the above works, you should save about 8 cycles on this (`pushw` and `popw` instructions are more expensive as compared to `loadw` and `storew`).

Reply:
Just a small modification and it starts working as expected. And it's cheaper. 

```bash
storew.local.0

movup.9
loadw.mem
movup.8
pushw.mem
```